neo4j/session_with_context.go

```diff
@@ -545,49 +583,126 @@ func (s *sessionWithContext) getServers(mode idb.AccessMode) func() []string {
-		s.log.Debugf(log.Session, s.logId, "connection acquisition timeout is %s, resolved deadline is: %s", timeout, deadline)
-	} else if deadline, ok := ctx.Deadline(); ok {
-		s.log.Debugf(log.Session, s.logId, "connection acquisition user-provided deadline is: %s", deadline)
 	}
 
-	if err := s.resolveHomeDatabase(ctx); err != nil {
-		return nil, errorutil.WrapError(err)
-	}
-	_, err := s.getOrUpdateServers(ctx, mode)
+	serverList, usedHomeDbGuess, err := s.getServerList(ctx, mode)
 	if err != nil {
 		return nil, errorutil.WrapError(err)
 	}
 
-	conn, err := s.pool.Borrow(
-		ctx,
-		s.getServers(mode),
-		timeout != 0,
-		s.config.BoltLogger,
-		livenessCheckTimeout,
-		s.auth)
+	conn, err := s.borrowConnection(ctx, serverList, livenessCheckTimeout)
 	if err != nil {
 		return nil, errorutil.WrapError(err)
 	}
 
+	// Return the connection if SSR is disabled, and we used the home database guess, then resolve as normal.
+	if !conn.IsSsrEnabled() && usedHomeDbGuess {
+		s.log.Debugf(
+			log.Session,
+			s.logId,
+			"Connection without SSR enabled was acquired using home database guess '%s', "+
+				"falling back to resolving as normal",
+			s.homeDbGuess,
+		)
+		s.pool.Return(ctx, conn)
+
+		serverList, _, err = s.getServerList(ctx, mode)
+		if err != nil {
+			return nil, errorutil.WrapError(err)
+		}
+
+		conn, err = s.borrowConnection(ctx, serverList, livenessCheckTimeout)
+		if err != nil {
+			return nil, errorutil.WrapError(err)
+		}
+	}
+
+	conn.SetPinHomeDatabaseCallback(s.pinHomeDatabaseCallback)
+
 	// Select database on server
 	if s.config.DatabaseName != idb.DefaultDatabase {
-		dbSelector, ok := conn.(idb.DatabaseSelector)
-		if !ok {
+		if err := s.selectDatabase(ctx, conn); err != nil {
 			s.pool.Return(ctx, conn)
-			return nil, &UsageError{Message: "Database does not support multi-database"}
+			return nil, err
 		}
-		dbSelector.SelectDatabase(s.config.DatabaseName)
 	}
 
 	return conn, nil
 }
 
+// applyConnectionTimeout sets a timeout on the context if configured.
+func (s *sessionWithContext) applyConnectionTimeout(ctx context.Context) (context.Context, context.CancelFunc) {
+	timeout := s.driverConfig.ConnectionAcquisitionTimeout
+	if timeout > 0 {
+		ctx, cancel := context.WithTimeout(ctx, timeout)
+		deadline, _ := ctx.Deadline()
+		s.log.Debugf(log.Session, s.logId, "connection acquisition timeout is %s, resolved deadline is: %s", timeout, deadline)
+		return ctx, cancel
+	}
+	if deadline, ok := ctx.Deadline(); ok {
+		s.log.Debugf(log.Session, s.logId, "connection acquisition user-provided deadline is: %s", deadline)
+	}
+	return ctx, nil
+}
+
+// getServerList resolves the server list based on the session configuration.
+// It returns a list of servers, a boolean indicating whether the home database guess was used, and an error if resolution fails.
+func (s *sessionWithContext) getServerList(ctx context.Context, mode idb.AccessMode) ([]string, bool, error) {
+	if s.config.DatabaseName != "" {
```

Comment:
```suggestion
	if !s.resolveHomeDb {
```
Should be equivalent, but 
 * faster when either connected to old servers that do not resolve the home db (pre 4.3, I believe) or when using a direct driver (`bolt+[s[sc]]://...`) where the home db resolution is a no-op.
 * avoids logging `"No valid cached home database guess available, resolving home database"` while the resolution will be a no-op.

Reply:
yeah nice change, can confirm tests still pass too